src/lib/npfg/npfg.hpp

```diff
@@ -0,0 +1,743 @@
+	 * Navigate on a fixed heading.
+	 *
+	 * This only holds a certain (air mass relative) direction and does not perform
+	 * cross track correction. Helpful for semi-autonomous modes. Introduced
+	 * by in ECL_L1_Pos_Controller, augmented for use with NPFG here.
+	 *
+	 * @param[in] heading_ref Reference heading angle [rad]
+	 * @param[in] ground_vel Vehicle ground velocity vector [m/s]
+	 * @param[in] wind_vel Wind velocity vector [m/s]
+	 */
+	void navigateHeading(float heading_ref, const matrix::Vector2f &ground_vel,
+			     const matrix::Vector2f &wind_vel);
+
+	/*
+	 * Navigate on a fixed bearing.
+	 *
+	 * This only holds a certain (ground relative) direction and does not perform
+	 * cross track correction. Helpful for semi-autonomous modes. Similar to navigateHeading.
+	 *
+	 * @param[in] bearing Bearing angle [rad]
+	 * @param[in] ground_vel Vehicle ground velocity vector [m/s]
+	 * @param[in] wind_vel Wind velocity vector [m/s]
+	 */
+	void navigateBearing(float bearing, const matrix::Vector2f &ground_vel, const matrix::Vector2f &wind_vel);
+
+	/*
+	 * Keep the wings level.
+	 *
+	 * @param[in] heading Heading angle [rad]
+	 */
+	void navigateLevelFlight(const float heading);
+
+	/*
+	 * [Copied directly from ECL_L1_Pos_Controller]
+	 *
+	 * Set the maximum roll angle output in radians
+	 */
+	void setRollLimit(float roll_lim_rad) { roll_lim_rad_ = roll_lim_rad; }
+
+	/*
+	 * [Copied directly from ECL_L1_Pos_Controller]
+	 *
+	 * Set roll angle slew rate. Set to zero to deactivate.
+	 */
+	void setRollSlewRate(float roll_slew_rate) { roll_slew_rate_ = roll_slew_rate; }
+
+	/*
+	 * [Copied directly from ECL_L1_Pos_Controller]
+	 *
+	 * Set control loop dt. The value will be used to apply roll angle setpoint slew rate limiting.
+	 */
+	void setDt(const float dt) { dt_ = dt; }
+
+	/*
+	 * [Copied directly from ECL_L1_Pos_Controller]
+	 *
+	 * Get the switch distance
+	 *
+	 * This is the distance at which the system will switch to the next waypoint.
+	 * This depends on the period and damping
+	 *
+	 * @param[in] wp_radius The switching radius the waypoint has set.
+	 */
+	float switchDistance(float wp_radius) const;
+
+	/*
+	 * The path bearing
+	 *
+	 * @return bearing angle (-pi..pi, in NED frame) [rad]
+	 */
+	float targetBearing() const { return atan2f(unit_path_tangent_(1), unit_path_tangent_(0)); }
+
+	/*
+	 * [Copied directly from ECL_L1_Pos_Controller]
+	 *
+	 * Returns true if the loiter waypoint has been reached
+	 */
+	bool reachedLoiterTarget() { return circleMode(); }
+
+	/*
+	 * Returns true if following a circle (loiter)
+	 */
+	bool circleMode() { return path_type_loiter_ && track_proximity_ > NPFG_EPSILON; }
+
+	/*
+	 * [Copied directly from ECL_L1_Pos_Controller]
+	 *
+	 * Get roll angle setpoint for fixed wing.
+	 *
+	 * @return Roll angle (in NED frame)
+	 */
+	float getRollSetpoint() { return roll_setpoint_; }
+
+private:
+
+	static constexpr float NPFG_EPSILON = 1.0e-4; // small number *bigger than machine epsilon
```

Comment:
Just for my curiosity: why is a custom epsilon needed here?

Reply:
Two main reasons: 1) it's used for WP distance thresholding (for if they are considered one point or not), which machine epsilon may miss if there is any floating point precision errors when we do the double to float conversions (global to local). 2) I'm (perhaps unreasonably) paranoid about dividing by very very small numbers, which could happen with how this epsilon is used in other places. Machine epsilon is very very small.

Reply:
Makes sense, thanks for the background!